### PR TITLE
CODAP-1384: fix Rescale to include all points across the antimeridian

### PR DESCRIPTION
**Pull request by bfinzer**
## Summary
- Rewrite `getRationalLongitudeBounds` as a gap-based algorithm (largest cyclic gap → complement arc) so Rescale no longer drops points across the date line. Inputs outside [-180, 180] are canonicalized; a compact dateline-crossing arc is returned with `max > 180` when appropriate.
- Mirror CODAP V2's per-point shift: introduce `shiftLongitudeIntoView` and apply it at every (lat, long) → container-point projection in the point and pin layers, so canonical-longitude points render inside a viewport whose east bound exceeds 180°.
- Cap longitude expansion in `expandLatLngBounds` at 360° to keep `fitBounds` well-defined for near-global views.

Fixes CODAP-1384.

## Known follow-up
- Polygon layers still position via Leaflet's `L.Polygon` at canonical longitudes; for a polygon dataset clustered across the date line, polygons in the canonical world copy could render off-screen in a dateline-crossing view. CODAP-1384's repro has no polygons; worth a separate ticket.
- `getRationalLongitudeBounds` no longer mutates the input array. The only in-tree caller already passed a copy.

## Test plan
- [x] `npm test -- src/components/map` — 101 tests pass, including new coverage of the CODAP-1384 dataset shape, tight Hawaii↔Russia clusters, out-of-range input canonicalization, multi-rotation shifts, and the "viewport too narrow to reach" edge case.
- [x] `npm run lint`
- [x] `npm run build:tsc`
- [x] Manual: opened `Map rescale problem.codap`. Initial map and Rescale fit every point.
- [x] Manual: pared-down dataset with only the dateline-clustered points — Rescale produces a compact Hawaii↔Russia viewport with all points rendered correctly.

### CODE REVIEW

File: v3/src/components/map/utilities/map-utils.ts [hunk 1]
```diff
@@ -13,65 +13,81 @@ export const isBoundaryAttribute = (attribute: IAttribute) => {
...
-    }
-    tLength = longs.length
-    if (tMax < tMin) {
-      const tTemp = tMax
-      tMax = tMin
-      tMin = tTemp
+  const sorted: number[] = []
+  for (const lng of longs) {
+    if (isFiniteNumber(lng)) sorted.push(canonicalizeLongitude(lng))
+  }
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/map/utilities/map-utils.ts:49]: Good catch! Added an O(n) fast path that tracks the canonical min/max and returns directly when the span is within 180°, skipping the sort. This also matches V2, which only ran the date-line algorithm when the span exceeded 180°.

File: v3/src/components/map/components/map-point-layer.tsx [hunk 2]
```diff
@@ -61,10 +62,14 @@ export const MapPointLayer = observer(function MapPointLayer({mapLayerModel, lay
     const {latId, longId} = mapLayerModel.pointAttributes || {}
     if (!dataset || !latId || !longId) return
 
+    const mapBounds = leafletMap.getBounds()
+    const west = mapBounds.getWest()
+    const east = mapBounds.getEast()
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/map/components/map-point-layer.tsx]: Good catch! Hoisted `getBounds()` into `connectingLinesForCases` so the viewport bounds are computed once per refresh and passed into the per-case projection.

File: v3/src/components/map/utilities/map-utils.test.ts [hunk 3]
```diff
@@ -74,4 +128,49 @@ describe("computeBoundsFromCoordinates", () => {
...
+    // outside the viewport so the point renders off-screen.
+    const result = shiftLongitudeIntoView(0, 175, 201)
+    expect(result < 175 || result > 201).toBe(true)
+  })
+
+  it("applies multiple rotations when a single 360° shift is not enough", () => {
+    // A point at -500° is two full rotations west of the viewport.
+    expect(shiftLongitudeIntoView(-500, 175, 201)).toBe(220)
+    // And a point at +700° is two rotations east.
+    expect(shiftLongitudeIntoView(700, -185, -159)).toBe(-380)
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/map/utilities/map-utils.test.ts:174]: Good catch! Updated the comment to "three rotations east (700 - 3×360 = -380)" — the assertion value was already correct.

**Review by kswenson — APPROVED** [review]
👍 Looks good -- the following review was developed in conjunction with Claude Code 🤖.

## PR Review Summary — #2618

**Changes:** 4 files, +203 / -68 lines

### What it does

Fixes CODAP-1384: "Rescale" on a map would drop points whose longitudes straddle the international date line (e.g. a dataset clustered around Hawaii + Russia). The old `getRationalLongitudeBounds` used an iterative shift-and-pop heuristic that could compute a bounding box that excluded some points, so Rescale didn't fit them all.

### The approach

Three coordinated changes:

1. **`getRationalLongitudeBounds` rewritten as a gap-based algorithm** (map-utils.ts). Canonicalize all longitudes to [-180, 180], sort, find the largest cyclic gap between consecutive points; the smallest enclosing arc is the *complement* of that gap. If the largest gap is the wrap gap (easternmost → westernmost across the date line), return the simple sorted min/max; otherwise return a date-line-crossing arc whose `max` exceeds 180°. This is the textbook "smallest arc covering points on a circle" solution and is provably correct, unlike the prior heuristic.

2. **Per-point longitude shift** (`shiftLongitudeIntoView`, applied in map-point-layer and map-pin-layer). When the fitted viewport crosses the date line (east > 180), a point stored at canonical longitude must be rendered in the adjacent world copy. `shiftLongitudeIntoView` adds/subtracts whole 360° rotations to bring the point into [west, east]. Mirrors CODAP V2 behavior. Applied at every (lat, long) → container projection: points, pins, heatmap, and connecting lines.

3. **`expandLatLngBounds` caps longitude expansion at 360°** so padding a near-global arc doesn't wrap past itself and confuse Leaflet's `fitBounds`.

`computeBoundsFromCoordinates` now always routes through `getRationalLongitudeBounds` (previously only when span > 180°) and adds proper non-finite guards.

### Assessment

The approach is sound and the algorithm is a clear improvement over the heuristic it replaces — both more correct and easier to reason about. Reactivity is preserved: each render/refresh path reads `getBounds()` fresh (point refresh is driven by the existing center/zoom reaction; pins/connecting-lines recompute on re-render), so points re-shift correctly on pan/zoom. All callers were checked — `map-content-model.rescale` and `use-map-model` consume the bounds via `expandLatLngBounds`, and the polygon vertex-collection path is unaffected.

Test coverage is strong: 23 tests pass, including the CODAP-1384 dataset shape, tight Hawaii↔Russia clusters, out-of-range canonicalization, multi-rotation shifts, non-finite filtering, single-value, and the "viewport too narrow to reach" edge case.

### V2 comparison

Checked against V2 on `master`. V2 has two date-line mechanisms:

- **Per-point shift** (`map_point_layer.js:191-196`): `shiftLongitudeIntoView` is a faithful line-for-line port. v3 had been *missing* this shift, which is a primary reason points dropped. This PR **restores V2 feature parity.** ✅
- **Bounds algorithm** (`map_point_data_configuration.js:195`): V2 uses the same median-based shift-and-pop heuristic the pre-PR v3 code was ported from. This PR **replaces** it with the gap-based algorithm — a deliberate **improvement** over V2, not a regression. Acceptable per the v3 goal (V2 parity as the floor, improvement allowed).

### Issues

No blocking issues. Minor observations (non-blocking, author's call:

- **Polygon layers across the date line** — the author documents this as a known follow-up; the CODAP-1384 repro has no polygons. Confirmed against V2 on : V2 renders polygons via  at canonical longitudes with **no per-vertex shift** (the shift existed only in V2's point layer, never the polygon layer). So this is a V2-parity *match*, not a parity gap — a genuine improvement opportunity for a future ticket, and correctly out of scope here.
- **Initial-frame ** — if Leaflet hasn't sized the map yet, an early refresh could read degenerate bounds and shift a point by 360°; this self-corrects on the next refresh (layout/zoom reaction). Pre-existing projection behavior, transient at most.

### Changes made during review (Copilot-inspired)

Copilot's review surfaced three minor items; all were addressed on the branch (commits  and ) and verified (tsc clean, 101/101 map tests, lint clean):

1. **Sort fast-path** (, map-utils.ts) — track the canonical min/max during the existing O(n) filter pass and early-return  without sorting when the span is within 180°. Correct because a ≤180° span guarantees the wrap gap exceeds every interior gap. Also restores V2's  gate, which the rewrite had dropped.
2. **Hoist ** (map-point-layer.tsx) —  ran once per case and recomputed the viewport bounds each time; now computed once per refresh in  and passed in, removing N−1 redundant calls.
3. **Test comment accuracy** (map-utils.test.ts) — the +700° multi-rotation case is three rotations (700 − 3×360 = −380), not two as the comment stated; assertion value was already correct.

All three Copilot threads resolved across two review rounds.

Recommendation: looks good to approve.
EOF
)

**Review by kswenson — APPROVED** [review]
👍 Looks good -- the following review was developed in conjunction with Claude Code 🤖.

## PR Review Summary — #2618

**Changes:** 4 files, +203 / -68 lines

### What it does

Fixes CODAP-1384: "Rescale" on a map would drop points whose longitudes straddle the international date line (e.g. a dataset clustered around Hawaii + Russia). The old `getRationalLongitudeBounds` used an iterative shift-and-pop heuristic that could compute a bounding box that excluded some points, so Rescale didn't fit them all.

### The approach

Three coordinated changes:

1. **`getRationalLongitudeBounds` rewritten as a gap-based algorithm** (map-utils.ts). Canonicalize all longitudes to [-180, 180], sort, find the largest cyclic gap between consecutive points; the smallest enclosing arc is the *complement* of that gap. If the largest gap is the wrap gap (easternmost → westernmost across the date line), return the simple sorted min/max; otherwise return a date-line-crossing arc whose `max` exceeds 180°. This is the textbook "smallest arc covering points on a circle" solution and is provably correct, unlike the prior heuristic.

2. **Per-point longitude shift** (`shiftLongitudeIntoView`, applied in map-point-layer and map-pin-layer). When the fitted viewport crosses the date line (east > 180), a point stored at canonical longitude must be rendered in the adjacent world copy. `shiftLongitudeIntoView` adds/subtracts whole 360° rotations to bring the point into [west, east]. Mirrors CODAP V2 behavior. Applied at every (lat, long) → container projection: points, pins, heatmap, and connecting lines.

3. **`expandLatLngBounds` caps longitude expansion at 360°** so padding a near-global arc doesn't wrap past itself and confuse Leaflet's `fitBounds`.

`computeBoundsFromCoordinates` now always routes through `getRationalLongitudeBounds` (previously only when span > 180°) and adds proper non-finite guards.

### Assessment

The approach is sound and the algorithm is a clear improvement over the heuristic it replaces — both more correct and easier to reason about. Reactivity is preserved: each render/refresh path reads `getBounds()` fresh (point refresh is driven by the existing center/zoom reaction; pins/connecting-lines recompute on re-render), so points re-shift correctly on pan/zoom. All callers were checked — `map-content-model.rescale` and `use-map-model` consume the bounds via `expandLatLngBounds`, and the polygon vertex-collection path is unaffected.

Test coverage is strong: 23 tests pass, including the CODAP-1384 dataset shape, tight Hawaii↔Russia clusters, out-of-range canonicalization, multi-rotation shifts, non-finite filtering, single-value, and the "viewport too narrow to reach" edge case.

### V2 comparison

Checked against V2 on `master`. V2 has two date-line mechanisms:

- **Per-point shift** (`map_point_layer.js:191-196`): `shiftLongitudeIntoView` is a faithful line-for-line port. v3 had been *missing* this shift, which is a primary reason points dropped. This PR **restores V2 feature parity.** ✅
- **Bounds algorithm** (`map_point_data_configuration.js:195`): V2 uses the same median-based shift-and-pop heuristic the pre-PR v3 code was ported from. This PR **replaces** it with the gap-based algorithm — a deliberate **improvement** over V2, not a regression. Acceptable per the v3 goal (V2 parity as the floor, improvement allowed).

### Issues

No blocking issues. Minor observations (non-blocking, author's call):

- **Polygon layers across the date line** — the author documents this as a known follow-up; the CODAP-1384 repro has no polygons. Confirmed against V2 on `master`: V2 renders polygons via `L.geoJson` at canonical longitudes with **no per-vertex shift** (the shift existed only in V2's point layer, never the polygon layer). So this is a V2-parity *match*, not a parity gap — a genuine improvement opportunity for a future ticket, and correctly out of scope here.
- **Initial-frame `getBounds()`** — if Leaflet hasn't sized the map yet, an early refresh could read degenerate bounds and shift a point by 360°; this self-corrects on the next refresh (layout/zoom reaction). Pre-existing projection behavior, transient at most.

### Changes made during review (Copilot-inspired)

Copilot's review surfaced three minor items; all were addressed on the branch (commits `85a7456c3` and `1d31a0686`) and verified (tsc clean, 101/101 map tests, lint clean):

1. **Sort fast-path** (`getRationalLongitudeBounds`, map-utils.ts) — track the canonical min/max during the existing O(n) filter pass and early-return `{min, max}` without sorting when the span is within 180°. Correct because a ≤180° span guarantees the wrap gap exceeds every interior gap. Also restores V2's `>180°` gate, which the rewrite had dropped.
2. **Hoist `getBounds()`** (map-point-layer.tsx) — `connectingLine()` ran once per case and recomputed the viewport bounds each time; now computed once per refresh in `connectingLinesForCases` and passed in, removing N−1 redundant calls.
3. **Test comment accuracy** (map-utils.test.ts) — the +700° multi-rotation case is three rotations (700 − 3×360 = −380), not two as the comment stated; assertion value was already correct.

All three Copilot threads resolved across two review rounds.

Recommendation: looks good to approve.
